### PR TITLE
[minor] add a general FP8ScaleSweepCalibrator and its registry

### PR DESCRIPTION
**Pull request by Fridah-nv**
### What does this PR do?

Type of change: ? 

### Usage

```python
# Add a code snippet demonstrating how to use this
```

### Before your PR is "*Ready for review*"

Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md) and your commits are signed (`git commit -s -S`).

Make sure you read and follow the [Security Best Practices](https://github.com/NVIDIA/Model-Optimizer/blob/main/SECURITY.md#security-coding-practices-for-contributors) (e.g. avoiding hardcoded `trust_remote_code=True`, `torch.load(..., weights_only=False)`, `pickle`, etc.).

- Is this change backward compatible?: ✅ / ❌ / N/A 
- If you copied code from any other sources or added a new PIP dependency, did you follow guidance in `CONTRIBUTING.md`: ✅ / ❌ / N/A 
- Did you write any new necessary tests?: ✅ / ❌ / N/A 
- Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?: ✅ / ❌ / N/A 

## Summary by CodeRabbit

* **New Features**
  * Added FP8 scale sweep calibration option for weight quantizers across standard and Hessian-based calibration pipelines.
  * Introduced a public API for registering custom FP8 scale sweep calibrators per backend, enabling extensibility.

* **Refactor**
  * Unified calibrator class structure to centralize FP8 scale sweep candidate generation logic.

### CODE REVIEW

**Review by cjluo-nv — COMMENTED** [review]
The refactor to extract `FP8ScaleSweepCalibrator` from `NVFP4MSECalibrator` is clean and the registry pattern for extending FP8 sweep calibration to other backends is reasonable. However, there are a few issues:

1. **Missing tests**: No tests for the new `FP8ScaleSweepCalibrator` class used directly, no tests for `register_fp8_sweep_calibrator`, and no tests for the registry-based code path in `mse_calibrate`/`local_hessian_calibrate`. The existing tests only cover `NVFP4MSECalibrator`.

2. **No validation in registry function**: `register_fp8_sweep_calibrator` accepts any `type` without checking it's actually a `FP8ScaleSweepCalibrator` subclass or has the expected constructor signature.

3. **Duplicated registry lookup logic**: The same pattern (compute `is_fp8_static_per_block_scales`, then if/else for NVFP4 vs. registry) is duplicated in both `mse_calibrate` and `local_hessian_calibrate`.

File: modelopt/torch/quantization/model_calib.py [hunk 1]
```diff
@@ -54,11 +54,31 @@
...
+    whose ``backend`` attribute matches a registered key will use the corresponding
+    calibrator class instead of the default :class:`MseCalibrator`.
+
+    Args:
+        backend: Backend name string (must match ``TensorQuantizer.backend``).
+        calibrator_cls: A :class:`FP8ScaleSweepCalibrator` subclass whose constructor
+            accepts ``(amax, axis, quant_func)``.
+    """
+    _FP8_SWEEP_CALIBRATOR_REGISTRY[backend] = calibrator_cls
+
```
Comment (cjluo-nv) [inline modelopt/torch/quantization/model_calib.py:81]: No validation that `calibrator_cls` is a subclass of `FP8ScaleSweepCalibrator` or that its constructor matches the expected signature `(amax, axis, quant_func, error_func)`. Consider adding a check:
```python
from .calib.mse import FP8ScaleSweepCalibrator
if not (isinstance(calibrator_cls, type) and issubclass(calibrator_cls, FP8ScaleSweepCalibrator)):
    raise TypeError(
        f"calibrator_cls must be a subclass of FP8ScaleSweepCalibrator, got {calibrator_cls}"
    )
```
This would catch registration errors early rather than at calibration time.

File: modelopt/torch/quantization/model_calib.py [hunk 2]
```diff
@@ -328,21 +348,33 @@ def mse_calibrate(
                     and module._block_sizes.get("scale_bits") == (4, 3)
                 )
```
Comment (cjluo-nv) [inline modelopt/torch/quantization/model_calib.py:350]: The logic for `is_fp8_static_per_block_scales`, the NVFP4-vs-registry branching, and calibrator instantiation is duplicated almost identically in `local_hessian_calibrate` (around line 648). Consider extracting a helper function like:
```python
def _make_fp8_sweep_calibrator(quantizer, initial_amax, quant_func, error_func=None):
    ...
```
to reduce duplication and ensure both code paths stay in sync.

File: modelopt/torch/quantization/model_calib.py [hunk 3]
```diff
@@ -328,21 +348,33 @@ def mse_calibrate(
...
-                    module._calibrator = NVFP4MSECalibrator(
-                        amax=initial_amax,
-                        axis=module._calibrator._axis,
-                        global_amax=module.global_amax,
-                        quant_func=partial(_mse_quant_func, quantizer=module),
-                    )
+                if fp8_scale_sweep and is_fp8_static_per_block_scales:
+                    if is_nvfp4_static:
+                        module._calibrator = NVFP4MSECalibrator(
+                            amax=initial_amax,
```
Comment (cjluo-nv) [inline modelopt/torch/quantization/model_calib.py:366]: Potential `AttributeError`: if the module matches the registry (i.e., `getattr(module, 'backend', None) in _FP8_SWEEP_CALIBRATOR_REGISTRY` is True) but `module._calibrator` is `None`, then `module._calibrator._axis` will raise `AttributeError`. The `local_hessian_calibrate` path handles this with `weight_quantizer._calibrator._axis if weight_quantizer._calibrator else None`, but this `mse_calibrate` path does not. Consider adding the same guard:
```python
axis=module._calibrator._axis if module._calibrator else None,
```

File: modelopt/torch/quantization/calib/mse.py [hunk 4]
```diff
@@ -171,30 +171,49 @@ def compute_amax(self, verbose: bool = False):
...
 
 
-class NVFP4MSECalibrator(MseCalibrator):
-    """Per-block FP8 scale sweep calibrator for NVFP4 static quantization."""
+class FP8ScaleSweepCalibrator(MseCalibrator):
+    """MSE calibrator that sweeps 126 valid FP8 E4M3 candidates of ``initial_amax``.
+
+    Candidate amax values are ``initial_amax * candidate``
+    """
+
```
Comment (cjluo-nv) [inline modelopt/torch/quantization/calib/mse.py:179]: Nit: The docstring says "Candidate amax values are `initial_amax * candidate`" but could be more explicit about the range. It might be helpful to note that candidates range from approximately 0.002 to 1.0 (valid positive finite FP8 E4M3 values divided by 448), so this sweeps the amax downward from `initial_amax`.

File: modelopt/torch/quantization/model_calib.py [hunk 5]
```diff
@@ -54,11 +54,31 @@
...
     "sequential_calibrate",
     "smoothquant",
     "svdquant",
 ]
 
+# Registry for backends that want a custom FP8-sweep calibrator for mse_calibrate().
+# Keys are backend name strings; values are FP8ScaleSweepCalibrator (sub)classes.
+_FP8_SWEEP_CALIBRATOR_REGISTRY: dict[str, type] = {}
+
+
```
Comment (cjluo-nv) [inline modelopt/torch/quantization/model_calib.py:67]: Missing tests: There are no tests for `register_fp8_sweep_calibrator`, the `_FP8_SWEEP_CALIBRATOR_REGISTRY` lookup path, or `FP8ScaleSweepCalibrator` used independently (without the NVFP4 `global_amax`). Please add tests covering:
1. Registering a custom calibrator and verifying it's used during `mse_calibrate`/`local_hessian_calibrate`
2. `FP8ScaleSweepCalibrator` collect/compute_amax workflow
3. Edge cases like registering a backend that doesn't exist on any quantizer
